core/src/main/java/hudson/model/FileFingerprintStorage.java

```diff
@@ -0,0 +1,170 @@
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Extension
+public class FileFingerprintStorage extends FingerprintStorage {
+
+    private static final Logger logger = Logger.getLogger(Fingerprint.class.getName());
+    private static final DateConverter DATE_CONVERTER = new DateConverter();
+
+    public @CheckForNull Fingerprint load(@NonNull byte[] md5sum) throws IOException {
+        return load(Fingerprint.getFingerprintFile(md5sum));
+    }
+
+    static @CheckForNull Fingerprint load(@NonNull File file) throws IOException {
+        XmlFile configFile = Fingerprint.getConfigFile(file);
+        if(!configFile.exists())
+            return null;
+
+        long start=0;
+        if(logger.isLoggable(Level.FINE))
+            start = System.currentTimeMillis();
+
+        try {
+            Object loaded = configFile.read();
+            if (!(loaded instanceof Fingerprint)) {
+                throw new IOException("Unexpected Fingerprint type. Expected " + Fingerprint.class + " or subclass but got "
+                        + (loaded != null ? loaded.getClass() : "null"));
+            }
+            Fingerprint f = (Fingerprint) loaded;
+            if(logger.isLoggable(Level.FINE))
+                logger.fine("Loading fingerprint "+file+" took "+(System.currentTimeMillis()-start)+"ms");
+            if (f.facets==null)
+                f.facets = new PersistedList<>(f);
+            for (FingerprintFacet facet : f.facets)
+                facet._setOwner(f);
+            return f;
+        } catch (IOException e) {
+            if(file.exists() && file.length()==0) {
+                // Despite the use of AtomicFile, there are reports indicating that people often see
+                // empty XML file, presumably either due to file system corruption (perhaps by sudden
+                // power loss, etc.) or abnormal program termination.
+                // generally we don't want to wipe out user data just because we can't load it,
+                // but if the file size is 0, which is what's reported in HUDSON-2012, then it seems
+                // like recovering it silently by deleting the file is not a bad idea.
+                logger.log(Level.WARNING, "Size zero fingerprint. Disk corruption? {0}", configFile);
+                file.delete();
+                return null;
+            }
+            String parseError = Fingerprint.messageOfParseException(e);
+            if (parseError != null) {
+                logger.log(Level.WARNING, "Malformed XML in {0}: {1}", new Object[] {configFile, parseError});
+                file.delete();
+                return null;
+            }
+            logger.log(Level.WARNING, "Failed to load "+configFile,e);
+            throw e;
+        }
+    }
+
+    public synchronized void save(Fingerprint fp) throws IOException {
+        if(BulkChange.contains(fp))
```

Comment:
Maybe something to leave within the abstraction. It is hard to expect the implementations to handle it correctly